File: crates/bridge_rpc/src/bridge_v2/id.rs
use std::{
    fmt::Display,
    process,
    sync::{
        LazyLock,
        atomic::{AtomicU32, Ordering},
    },
};

use rand::Rng;

#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    Hash,
    serde::Serialize,
    serde::Deserialize,
    Default,
)]
#[serde(transparent)]
#[repr(transparent)]
pub struct Id(u64);

impl Id {
    pub fn new() -> Self {
        Self(unique_u64_for_pid())
    }
}

impl Display for Id {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

static COUNTER: LazyLock<AtomicU32> = LazyLock::new(|| {
    // seed counter with random start to reduce chance of collision after PID reuse
    let seed: u32 = rand::rng().random();
    AtomicU32::new(seed)
});

/// Returns a u64 that is unique per-call within the process and unique across
/// concurrently running processes on the same machine (PID in high 32 bits).
pub fn unique_u64_for_pid() -> u64 {
    let pid = process::id(); // u32
    let low = COUNTER.fetch_add(1, Ordering::Relaxed);
    ((pid as u64) << 32) | (low as u64)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    #[test]
    fn unique_in_process() {
        let mut seen = HashSet::new();
        for _ in 0..1_000_000u32 {
            let v = unique_u64_for_pid();
            assert!(seen.insert(v));
        }
    }
}

File: packages/bridge-rpc/src/id.ts
import * as z from "zod";

/**
 * A unique identifier that combines process ID with an incrementing counter.
 * Format: high 32 bits = PID, low 32 bits = counter
 */
export class Id {
    private readonly value: bigint;
    private constructor(value: bigint | number | string) {
        if (typeof value === "bigint") {
            this.value = value;
        } else {
            this.value = BigInt(value);
        }
    }

    /**
     * Generate a new unique ID for this process
     */
    static create(): Id {
        return new Id(uniqueU64ForPid());
    }

    /**
     * Create an Id from a string representation
     */
    static fromString(str: string): Id {
        return new Id(str);
    }

    static fromBigInt(value: bigint): Id {
        return new Id(value);
    }

    static fromNumber(value: number): Id {
        return new Id(value);
    }

    /**
     * Get the underlying u64 value as a bigint
     */
    public getValue(): bigint {
        return this.value;
    }

    public toNumber(): number {
        return Number(this.value);
    }

    public valueOf(): bigint {
        return this.value;
    }

    /**
     * Convert to string representation
     */
    public toString(): string {
        return this.value.toString();
    }

    /**
     * Convert to JSON (as number)
     */
    public toJSON(): number {
        return Number(this.value);
    }

    /**
     * Check equality with another Id
     */
    public equals(other: Id): boolean {
        return this.value === other.value;
    }
}

// Zod schema for Id
export const IdSchema = z.bigint().transform((b) => Id.fromBigInt(b));

// Global counter state
class Counter {
    private value: number;

    constructor() {
        // Seed counter with random start to reduce chance of collision after PID reuse
        this.value = Math.floor(Math.random() * 0xffffffff);
    }

    /**
     * Atomically increment and return the previous value
     * Note: JavaScript is single-threaded, so this is naturally atomic
     */
    fetchAdd(): number {
        const current = this.value;
        // Wrap around at 32-bit boundary
        this.value = (this.value + 1) >>> 0;
        return current;
    }
}

// Lazy-initialized global counter
let counter: Counter | null = null;

function getCounter(): Counter {
    if (counter === null) {
        counter = new Counter();
    }
    return counter;
}

/**
 * Returns a u64 (as bigint) that is unique per-call within the process
 * and unique across concurrently running processes on the same machine
 * (PID in high 32 bits, counter in low 32 bits).
 */
export function uniqueU64ForPid(): bigint {
    const pid = process.pid; // number
    const low = getCounter().fetchAdd();

    // Combine: (pid << 32) | low
    const pidBig = BigInt(pid >>> 0); // Ensure unsigned 32-bit
    const lowBig = BigInt(low >>> 0); // Ensure unsigned 32-bit

    return (pidBig << 32n) | lowBig;
}

/**
 * Reset the counter (useful for testing)
 * @internal
 */
export function resetCounter(): void {
    counter = null;
}
